src/get_env.c

```diff
@@ -0,0 +1,60 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+#include "c_utilities/get_env.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#ifdef WIN32
+# define WINDOWS_ENV_BUFFER_SIZE 2048
+static char __env_buffer[WINDOWS_ENV_BUFFER_SIZE];
+#endif  // WIN32
+
+
+const char *
+utilities_get_env(const char * env_name, const char ** env_value)
+{
+  if (!env_name) {
+    return "argument env_name is null";
+  }
+  if (!env_value) {
+    return "argument env_value is null";
+  }
+  *env_value = NULL;
+#ifdef WIN32
+  size_t required_size;
+  errno_t ret = getenv_s(&required_size, __env_buffer, sizeof(__env_buffer), env_name);
```

Comment:
We could also provide a `utilities_dup_env()`. I was being conservative and trying to avoid memory allocation if possible. In cases where you know that the variable you're reading shouldn't be changing, like `ROS_DOMAIN_ID` I think you can safely use `utilities_get_env()`, but in other cases you might want the copy. On the other hand, `dupenv_s` is just for convenience, however, if does seem to behave differently from `get_env`:

> If the variable is not found, then buffer is set to NULL, numberOfElements is set to 0, and the return value is 0 because this situation is not considered to be an error condition.

-- https://msdn.microsoft.com/en-us/library/ms175774.aspx#Remarks

The `getenv_s` documentation is frustratingly unclear about this point, but I feel confident that I landed on the above behavior of `getenv_s` through experimentation and ultimately that test I pointed out.